sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/codegen/CodeGenerator.scala

```diff
@@ -1612,6 +1612,49 @@ object CodeGenerator extends Logging {
     }
   }
 
+  /**
+   * Extracts all the input variables from references and subexpression elimination states
+   * for a given `expr`. This result will be used to split the generated code of
+   * expressions into multiple functions.
+   */
+  def getLocalInputVariableValues(
+      context: CodegenContext,
+      expr: Expression,
+      subExprs: Map[Expression, SubExprEliminationState]): Seq[((String, DataType), Expression)] = {
+    val argMap = mutable.Map[(String, DataType), Expression]()
+    val stack = mutable.Stack[Expression](expr)
```

Comment:
Question: so the stack here is only used for implementing depth-first pre-order traversal of an expression tree in an iterative style instead of recursive style, right?

Reply:
Yea, right. I just followed the other similar logics.